File: JXBC-Server/appbase/src/JXBC.Web/JXBC.WebAPI/Controllers/PaymentController.cs
using System;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Web;
using System.Web.Http;
using M2SA.AppGenome;
using JXBC.TradeSystem;
using JXBC.WebCore;
using JXBC.WebCore.Mvc;
using JXBC.WebCore.ViewModels;
using M2SA.AppGenome.Cache;
using JXBC.Passports.Security;
using M2SA.AppGenome.Logging;
using JXBC.TradeSystem.Providers;

namespace JXBC.WebAPI.Controllers
{
    /// <summary>
    /// 
    /// </summary>
    public class PaymentController : AuthenticatedApiController
    {
        /// <summary>
        /// 
        /// </summary>
        /// <param name="bizSource"></param>
        /// <returns></returns>
        public AppleIAPProvider.Product GetIAPProduct(string bizSource)
        {            
            var iapProvider = (AppleIAPProvider)ObjectIOCFactory.GetSingleton<PaymentEngine>().PaymentProviders[PayWays.AppleIAP];
            return iapProvider.GetProductJson(bizSource);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="os"></param>
        /// <param name="bizSource"></param>
        /// <param name="payWays"></param>
        /// <returns></returns>
        [HttpGet]
        public string[] GetPayWays(string os, string bizSource, string payWays = null)
        {
            if(false == string.IsNullOrEmpty(payWays))
            {
                return payWays.Split(',');
            }

            return ObjectIOCFactory.GetSingleton<PaymentEngine>().GetPayWays(os, bizSource);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="payment"></param>
        /// <returns></returns>
        [HttpPost]
        public PaymentResult CreateTrade([FromBody]Payment payment)
        {
            payment.AssertNotNull("payment");
            if (string.IsNullOrEmpty(payment.BizSource)) throw new ArgumentNullException("payment:BizSource");
            if (string.IsNullOrEmpty(payment.CommoditySubject)) throw new ArgumentNullException("payment:CommoditySubject");

            if (payment.OwnerId == 0 && (payment.TradeType == TradeType.OrganizationToPersonal || payment.TradeType == TradeType.OrganizationToOrganization))
                throw new ArgumentOutOfRangeException("payment:OwnerId", string.Format("Not support OwnerId is 0 from OrganizationTo*** trade.", payment.TradeMode));

            if (string.IsNullOrEmpty(payment.PayRoute))
                payment.PayRoute = PayWays.Route_APP;

            payment.BuyerId = MvcContext.Current.PassportId;
            if (payment.TradeType == TradeType.PersonalToPersonal || payment.TradeType == TradeType.PersonalToOrganization)
            {
                payment.OwnerId = payment.BuyerId;
            }
            payment.ClientIP = "8.8.8.8";
            try
            {
                var paymentEngine = ObjectIOCFactory.GetSingleton<PaymentEngine>();
                var signed = paymentEngine.CreateTrade(payment);
                if (signed)
                {
                    return new PaymentResult() { Success = true, TradeCode = payment.TradeCode, SignedParams = payment.SignedParams };
                }
                else
                {
                    return new PaymentResult() { Success = false, ErrorMessage = "Create sign failed." };
                }
            }
            catch (ArgumentOutOfRangeException ex)
            {
                this.ReturnPreconditionFailedMessage(ex.Message);
                return null;
            }
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="paymentResult"></param>
        /// <returns></returns>
        [HttpPost]
        public PaymentResult PaymentCompleted([FromBody]PaymentResult paymentResult)
        {
            if (null == paymentResult || string.IsNullOrEmpty(paymentResult.TradeCode))
            {
                this.ReturnPreconditionFailedMessage();
                return null;
            }

            LogManager.GetLogger().Info("==>paymentResult[PaymentCompleted]: {0}", TradeSystem.JsonExtension.ToJson(paymentResult));

            var tradeJournal = ObjectIOCFactory.GetSingleton<PaymentEngine>().PaymentCompleted(paymentResult.TradeCode, paymentResult.PayWay, null, paymentResult.PaidDetail);
            if (null != tradeJournal)
                return new PaymentResult() { Success = true, TradeCode= tradeJournal.TradeCode, TargetBizTradeCode = tradeJournal.TargetBizTradeCode};
            else
                return new PaymentResult() { Success = false, ErrorMessage = "Paid failed." };
        }

        /// <summary>
        /// 
        /// </summary>
        /// <param name="tradeCode"></param>
        /// <returns></returns>
        [HttpGet]
        public PaymentResult QueryTrade(string tradeCode)
        {
            if (string.IsNullOrEmpty(tradeCode))
                return null;

            return ObjectIOCFactory.GetSingleton<PaymentEngine>().QueryTrade(tradeCode);
        }
    }
}

File: JXBC-Server/BossComments/apiserver/appbase/controllers/PaymentController.php
<?php

namespace app\appbase\controller;

/**
 * @SWG\Tag(
 *   name="Payment",
 *   description="支付相关API"
 * )
 */ 
class PaymentController
{
    /**
     * @SWG\Get(
     *     path="/appbase/Payment/getIAPProduct",
     *     summary="获取指定业务的苹果内购产品信息",
     *     tags={"Payment"},
     *     description="",
     *     @SWG\Parameter(
     *         name="bizSource",
     *         in="query",
     *         description="业务源",
     *         required=true,
     *         type="string"
     *     ),
     *     @SWG\Response(
     *         response=200,
     *         description="苹果内购产品信息",
     *         @SWG\Schema(
     *             ref="#/definitions/IAPProduct"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function getIAPProduct($request)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI");
    }

    /**
     * @SWG\Get(
     *     path="/appbase/Payment/getPayWays",
     *     summary="获取指定业务的所支持的支付方式",
     *     tags={"Payment"},
     *     description="",
     *     @SWG\Parameter(
     *         name="os",
     *         in="query",
     *         description="APP系统",
     *         required=true,
     *         type="string"
     *     ),
     *     @SWG\Parameter(
     *         name="bizSource",
     *         in="query",
     *         description="业务源",
     *         required=true,
     *         type="string"
     *     ),
     *     @SWG\Parameter(
     *         name="payWays",
     *         in="query",
     *         description="指定支付方式列表（仅供测试是使用）",
     *         required=false,
     *         type="string"
     *     ),
     *     @SWG\Response(
     *         response=200,
     *         description="支付方式列表",
     *         @SWG\Schema(
     *          type="array",
     *          @SWG\Items(type="string")
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function getPayWays($request)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI");
    }

    /**
     * @SWG\Definition(
     *   definition="PaymentResult:new",
     *   allOf={
     *     @SWG\Schema(ref="#/definitions/PaymentResult"),
     *     @SWG\Schema(required={"SignedParams"})
     *   },
     * )      
     * @SWG\POST(
     *     path="/appbase/Payment/createTrade",
     *     summary="创建交易记录",
     *     description="不同交易来源(BizSource)，使用商品相关参数组合<br/> <br/>对公交易时，必须设置参数<b>OwnerId</b>，开户时设置该值为 <b>-1</b>;",     
     *     tags={"Payment"},
     *     @SWG\Parameter(
     *         name="body",
     *         in="body",
     *         description="交易信息",
     *         required=true,
     *         @SWG\Schema(ref="#/definitions/Payment"),
     *     ),       
     *     @SWG\Response(
     *         response=200,
     *         description="交易记录信息，包括交易编码和第三方支付所需的（签名后）参数",
     *         @SWG\Schema(
     *             ref="#/definitions/PaymentResult:new"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(   
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function createTrade($entity)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }

    /**
     * @SWG\Definition(
     *   definition="PaymentResult:paid",
     *   allOf={
     *     @SWG\Schema(ref="#/definitions/PaymentResult"),
     *     @SWG\Schema(required={"PayWay","PaidDetail"})
     *   },
     * )    
     * @SWG\Post(
     *     path="/appbase/Payment/paymentCompleted",
     *     summary="处理APP支付结果，完成交易",
     *     tags={"Payment"},
     *     description="APP支付后，服务端进行后继处理(支付结果校验+修改交易状态+业务处理)",
     *     @SWG\Parameter(
     *         name="body",
     *         in="body",
     *         description="支付结果（包含第三方支付结果)",
     *         required=true,
     *         @SWG\Schema(ref="#/definitions/PaymentResult:paid"),
     *     ),     
     *     @SWG\Response(
     *         response=200,
     *         description="支付结果",
     *         @SWG\Schema(
     *             ref="#/definitions/PaymentResult"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(   
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function paymentCompleted($entity)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI"); 
    }

    /**
     * @SWG\Post(
     *     path="/appbase/Payment/queryTrade",
     *     summary="查询交易是否成功支付",
     *     tags={"Payment"},
     *     description="",
     *     @SWG\Parameter(
     *         name="tradeCode",
     *         in="query",
     *         description="交易编号",
     *         required=true,
     *         type="string"
     *     ),
     *     @SWG\Response(
     *         response=200,
     *         description="支付结果",
     *         @SWG\Schema(
     *             ref="#/definitions/PaymentResult"
     *         )
     *     ),
     *     @SWG\Response(
     *         response="412",
     *         description="不符合预期的输入参数",
     *         @SWG\Schema(
     *             ref="#/definitions/Error"
     *         )
     *     )
     * )
     */
    public function queryTrade($request)
    {
        throw new Exception("逻辑实现位于.net项目 JXBC.WebAPI");
    }
}
